src/test/common/test_random.cc

```diff
@@ -0,0 +1,272 @@
+  ceph::util::randomize_rng();
+ 
+  // Get a random int between 0 and max int:
+  auto a = ceph::util::generate_random_number();
+ 
+  // Get a random int between 0 and 20:
+  auto b = ceph::util::generate_random_number(20);
+ 
+  // Get a random int between 1 and 20:
+  auto c = ceph::util::generate_random_number(1, 20);
+ 
+  // Get a random float between 0.0 and 20.0:
+  auto d = ceph::util::generate_random_number(20.0);
+ 
+  // Get a random float between 0.001 and 0.991:
+  auto e = ceph::util::generate_random_number(0.001, 0.991);
+ 
+  // Make a random number function suitable for putting on its own thread:
+  auto gen_f = ceph::util::make_random_number_function<1, 20>();
+  auto x = gen_f();
+ 
+  // Same, with seed (no re-seeding is possible):
+  auto gen_g = ceph::util::make_random_number_function<1, 20, 42>();
+  auto y = gen_g();
+ 
+  // Make a function object RNG suitable for putting on its own thread:
+  auto gen_fn = ceph::util::random_number_generator<int>();
+  auto z = gen_fn();
+  gen_fn.seed(42);   // re-seed
+ 
+  // Mostly to swallow "unused variable" warnings:
+  ASSERT_EQ(true, none_equal(a, b, c, d, e, x, y, z));
+}
+
+TEST(util, test_random)
+{
+  /* The intent of this test is not to formally test random number generation, 
+  but rather to casually check that "it works" and catch regressions: */
+ 
+  // The default overload should compile:
+  ceph::util::randomize_rng();
+ 
+  {
+    int a = ceph::util::generate_random_number();
+    int b = ceph::util::generate_random_number();
+ 
+    /* Technically, this can still collide and cause a false negative, but let's 
+    be optimistic: */
+    if (std::numeric_limits<int>::max() > 32767) {
+       ASSERT_GT(a, -1);
+       ASSERT_GT(b, -1);
+ 
+       ASSERT_NE(a, b);
+     }
+  }
+ 
+  {
+  auto a = ceph::util::generate_random_number(1, std::numeric_limits<int>::max());
+  auto b = ceph::util::generate_random_number(1, std::numeric_limits<int>::max());
+ 
+  if (std::numeric_limits<int>::max() > 32767) {
+     ASSERT_GT(a, 0);
+     ASSERT_GT(b, 0);
+ 
+     ASSERT_NE(a, b);
+   }
+  }
+ 
+  for (auto n = 100000; n; --n) {
+     constexpr int min = 0, max = 6;
+     int a = ceph::util::generate_random_number<min, max>();
+     ASSERT_GT(a, -1);
+     ASSERT_LT(a, 7);
+   }
+ 
+  // Multiple types (integral):
+  {
+    int min = 0, max = 1;
+    type_check_ok(min, max);
+  }
+ 
+  {
+    long min = 0, max = 1l;
+    type_check_ok(min, max);
+  }
+ 
+  // Multiple types (floating point):
+  {
+    double min = 0.0, max = 1.0;
+    type_check_ok(min, max);
+  }
+ 
+  {
+    float min = 0.0, max = 1.0;
+    type_check_ok(min, max);
+  }
+ 
+  // min > max should not explode:
+  {
+    float min = 1.0, max = 0.0;
+    auto x = ceph::util::generate_random_number(min, max);
+  }
+}
+
+TEST(util, test_user_mutex)
+{
+  // Show that we can call with a user lock:
+  std::mutex l;
+ 
+  ceph::util::randomize_rng(l);
+ 
+  ceph::util::generate_random_number(l);
```

Comment:
i don't like how this part of the interface mixes global engines with caller-provided locks. this is error-prone, because it relies on the caller to provide the same lock for every call in order for it to actually be thread-safe. a global engine should be protected by a global mutex